In [2]:
!unzip /content/STAT+S01V22.zip

Archive:  /content/STAT+S01V22.zip
  inflating: bigmartsales.csv        
  inflating: Solution_for_Project_4.ipynb  


In [3]:
import pandas as pd
import numpy as np

In [4]:
sales_data = pd.read_csv("/content/bigmartsales.csv")
sales_data.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

**Calculate mean, median, mode for Item_MRP column.**

In [5]:
sales_data.columns

Index(['Item_Identifier', 'Item_Weight', 'Item_Fat_Content', 'Item_Visibility',
       'Item_Type', 'Item_MRP', 'Outlet_Identifier',
       'Outlet_Establishment_Year', 'Outlet_Size', 'Outlet_Location_Type',
       'Outlet_Type', 'Item_Outlet_Sales'],
      dtype='object')

In [6]:
print("Mean of Item_MRP column is:",sales_data['Item_MRP'].mean())
print("Median of Item_MRP column is:",sales_data['Item_MRP'].median())
print("Mode of Item_MRP column is:",sales_data['Item_MRP'].mode())

Mean of Item_MRP column is: 140.9927819781767
Median of Item_MRP column is: 143.0128
Mode of Item_MRP column is: 0    172.0422
Name: Item_MRP, dtype: float64


The above output concludes that:
- The average of the Item_MRP column is 140.99 ~ 141. Which means that the average MRP of items is about Rs.141/-
- The median of the Item_MRP column is 143 which means that approximately half of the items MRP is less than Rs.143/- and the other half is greater than Rs.143/-
- The mode of the column is 172 which means that the most frequently occurring item MRP is Rs.172/-
- As the mean and median values are almost equal, we can say that the distribution is almost symmetrical.

**Calculate the measure of spread - range, IQR, variance, std for Item_MRP column.**

In [7]:
from scipy.stats import iqr
print("Range of Item_MRP column is:",np.ptp(sales_data['Item_MRP']))
print("IQR of Item_MRP column is:",iqr(sales_data['Item_MRP']))
print("Variance of Item_MRP column is:",sales_data['Item_MRP'].var())
print("Standard Deviation of Item_MRP column is:",sales_data['Item_MRP'].std())

Range of Item_MRP column is: 235.5984
IQR of Item_MRP column is: 91.8172
Variance of Item_MRP column is: 3878.183909097746
Standard Deviation of Item_MRP column is: 62.27506651219046


The above output suggests that the Range of the Item_MRP column is vast which means that the item MRPs are varied. This is proved by calculating the variance of the column. The variance value is 3878 which indicates that the MRP of items si spread out vastly and has lot of variety.

**Obtain different summary statistics (skewness and kurtosis) for the Item_Outlet_Sales column.**

In [8]:
from scipy.stats import skew, kurtosis
print("Skewness of the Item_Outlet_Sales column is:",skew(sales_data['Item_Outlet_Sales']))
print(f"Pearsons kurtosis of the Item_Outlet_Sales column is: {kurtosis(sales_data['Item_Outlet_Sales'], fisher = False)}")
print(f"Fisher's kurtosis of Item_Outlet_Sales column is: {kurtosis(sales_data['Item_Outlet_Sales'], fisher = True)}")

Skewness of the Item_Outlet_Sales column is: 1.1773233539676373
Pearsons kurtosis of the Item_Outlet_Sales column is: 4.614224991590342
Fisher's kurtosis of Item_Outlet_Sales column is: 1.6142249915903424


The above output suggests that:
- As the skewness is 1.177, it indicates that the Item_Outlet_Sales column is positively skewed (right-skewed).
- As the Pearsons kurtosis value is 4.614 which is greater than 3 indicates a leptokurtic distribution i.e., a distribution with heavier tails and a sharper peak.

**Detect and remove outliers based on the data in Item_MRP and Item_Outlet_Sales columns using z-score.**

In [9]:
from scipy.stats import zscore
print("Z-score of Item_MRP and Item_Outlet_Sales column:")
z = abs(zscore(sales_data[['Item_MRP', 'Item_Outlet_Sales']], axis=0))
print(pd.DataFrame(z))

Z-score of Item_MRP and Item_Outlet_Sales column:
      Item_MRP  Item_Outlet_Sales
0     1.747454           0.910601
1     1.489023           1.018440
2     0.010040           0.049238
3     0.660050           0.849103
4     1.399220           0.695373
...        ...                ...
8518  1.180783           0.349915
8519  0.527301           0.956402
8520  0.897208           0.579100
8521  0.607977           0.196725
8522  1.052261           0.829594

[8523 rows x 2 columns]


In [13]:
outliers = np.where(z>3)
outliers

(array([ 130,  145,  276,  304,  333,  373,  456,  472,  661,  694,  753,
         809,  935,  991,  997, 1009, 1138, 1254, 1450, 1506, 1859, 1998,
        2109, 2216, 2238, 2293, 2516, 2535, 2776, 3083, 3087, 3098, 3154,
        3234, 3286, 3407, 3667, 3709, 3720, 3757, 4289, 4346, 4349, 4676,
        4681, 4772, 4888, 4961, 4991, 5065, 5177, 5185, 5223, 5315, 5500,
        5550, 5619, 5752, 5799, 5812, 5977, 6000, 6134, 6334, 6409, 6537,
        6541, 6606, 6886, 6993, 7016, 7076, 7188, 7191, 7236, 7256, 7302,
        7343, 7560, 7562, 7609, 7690, 7752, 7816, 7826, 7930, 8039, 8201,
        8245, 8447]),
 array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1]))

The above output indicates the row values which exceed the given threshold value of 3 and hence are considered as outliers.

In [28]:
np.size(outliers[0])

90

This output shows the total count of the outliers which we have to remove from the main data.

In [14]:
sales_data = sales_data.drop(outliers[0])

This piece of code is written to drop the outliers from the original data.

In [21]:
sales_data

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0              FDA15        9.300          Low Fat         0.016047   
1              DRC01        5.920          Regular         0.019278   
2              FDN15       17.500          Low Fat         0.016760   
3              FDX07       19.200          Regular         0.000000   
4              NCD19        8.930          Low Fat         0.000000   
...              ...          ...              ...              ...   
8518           FDF22        6.865          Low Fat         0.056783   
8519           FDS36        8.380          Regular         0.046982   
8520           NCJ29       10.600          Low Fat         0.035186   
8521           FDN46        7.210          Regular         0.145221   
8522           DRG01       14.800          Low Fat         0.044878   

                  Item_Type  Item_MRP Outlet_Identifier  \
0                     Dairy  249.8092            OUT049   
1               Soft Drinks   48.2692            OUT018   
2                      Meat  141.6180            OUT049   
3     Fruits and Vegetables  182.0950            OUT010   
4                 Household   53.8614            OUT013   
...                     ...       ...               ...   
8518            Snack Foods  214.5218            OUT013   
8519           Baking Goods  108.1570            OUT045   
8520     Health and Hygiene   85.1224            OUT035   
8521            Snack Foods  103.1332            OUT018   
8522            Soft Drinks   75.4670            OUT046   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                          1999      Medium               Tier 1   
1                          2009      Medium               Tier 3   
2                          1999      Medium               Tier 1   
3                          1998         NaN               Tier 3   
4                          1987        High               Tier 3   
...                         ...         ...                  ...   
8518                       1987        High               Tier 3   
8519                       2002         NaN               Tier 2   
8520                       2004       Small               Tier 2   
8521                       2009      Medium               Tier 3   
8522                       1997       Small               Tier 1   

            Outlet_Type  Item_Outlet_Sales  
0     Supermarket Type1          3735.1380  
1     Supermarket Type2           443.4228  
2     Supermarket Type1          2097.2700  
3         Grocery Store           732.3800  
4     Supermarket Type1           994.7052  
...                 ...                ...  
8518  Supermarket Type1          2778.3834  
8519  Supermarket Type1           549.2850  
8520  Supermarket Type1          1193.1136  
8521  Supermarket Type2          1845.5976  
8522  Supermarket Type1           765.6700  

[8433 rows x 12 columns]

**Detect outliers based on data in Item_Outlet_Sales column using IQR.**

In [32]:
Q1 = sales_data['Item_Outlet_Sales'].quantile(0.25)
Q3 = sales_data['Item_Outlet_Sales'].quantile(0.75)
IQR = Q3 - Q1
print("IQR of Item_Outlet_Sales column is:",IQR)
outlier_1 = sales_data['Item_Outlet_Sales'] < Q1 - 1.5*IQR
outlier_2 = sales_data['Item_Outlet_Sales'] > Q3 + 1.5*IQR
total_outliers = sales_data['Item_Outlet_Sales'][(outlier_1) | (outlier_2)]
print("Outliers based on IQR:")
print(total_outliers)

IQR of Item_Outlet_Sales column is: 2233.759
Outliers based on IQR:
43      6768.5228
132     6976.2524
203     6704.6060
240     6795.1548
243     7222.5984
          ...    
8113    6579.4356
8208    6615.3888
8329    6630.0364
8350    7240.5750
8510    7182.6504
Name: Item_Outlet_Sales, Length: 113, dtype: float64


The above output depicts that there are 113 rows which are considered as outliers based on the given formula. Hence, we need to drop them.

In [33]:
sales_data = sales_data.drop(total_outliers.index)
sales_data

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0              FDA15        9.300          Low Fat         0.016047   
1              DRC01        5.920          Regular         0.019278   
2              FDN15       17.500          Low Fat         0.016760   
3              FDX07       19.200          Regular         0.000000   
4              NCD19        8.930          Low Fat         0.000000   
...              ...          ...              ...              ...   
8518           FDF22        6.865          Low Fat         0.056783   
8519           FDS36        8.380          Regular         0.046982   
8520           NCJ29       10.600          Low Fat         0.035186   
8521           FDN46        7.210          Regular         0.145221   
8522           DRG01       14.800          Low Fat         0.044878   

                  Item_Type  Item_MRP Outlet_Identifier  \
0                     Dairy  249.8092            OUT049   
1               Soft Drinks   48.2692            OUT018   
2                      Meat  141.6180            OUT049   
3     Fruits and Vegetables  182.0950            OUT010   
4                 Household   53.8614            OUT013   
...                     ...       ...               ...   
8518            Snack Foods  214.5218            OUT013   
8519           Baking Goods  108.1570            OUT045   
8520     Health and Hygiene   85.1224            OUT035   
8521            Snack Foods  103.1332            OUT018   
8522            Soft Drinks   75.4670            OUT046   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                          1999      Medium               Tier 1   
1                          2009      Medium               Tier 3   
2                          1999      Medium               Tier 1   
3                          1998         NaN               Tier 3   
4                          1987        High               Tier 3   
...                         ...         ...                  ...   
8518                       1987        High               Tier 3   
8519                       2002         NaN               Tier 2   
8520                       2004       Small               Tier 2   
8521                       2009      Medium               Tier 3   
8522                       1997       Small               Tier 1   

            Outlet_Type  Item_Outlet_Sales  
0     Supermarket Type1          3735.1380  
1     Supermarket Type2           443.4228  
2     Supermarket Type1          2097.2700  
3         Grocery Store           732.3800  
4     Supermarket Type1           994.7052  
...                 ...                ...  
8518  Supermarket Type1          2778.3834  
8519  Supermarket Type1           549.2850  
8520  Supermarket Type1          1193.1136  
8521  Supermarket Type2          1845.5976  
8522  Supermarket Type1           765.6700  

[8320 rows x 12 columns]

The above output gives a total of 8320 rows as it has removed the 113 rows that are considered as outliers in the previous cells.